eventsources/sources/github/start.go

```diff
@@ -124,182 +125,183 @@ func (router *Router) HandleRoute(writer http.ResponseWriter, request *http.Requ
-		if _, err := router.githubClient.Repositories.DeleteHook(ctx, githubEventSource.Owner, githubEventSource.Repository, *router.hook.ID); err != nil {
-			return errors.Errorf("failed to delete hook. err: %+v", err)
+		logger.Info("setting up the upload url for GitHub client...")
+		if githubEventSource.GithubUploadURL != "" {
+			uploadURL, err := url.Parse(githubEventSource.GithubUploadURL)
+			if err != nil {
+				return errors.Errorf("failed to parse github upload url. err: %s", err)
+			}
+			client.UploadURL = uploadURL
 		}
-		logger.Info("GitHub hook deleted")
-	}
+		router.githubClient = client
+		router.hookIDs = make(map[string]int64)
 
-	return nil
-}
+		hook := &gh.Hook{
+			Events: githubEventSource.Events,
+			Active: gh.Bool(githubEventSource.Active),
+			Config: hookConfig,
+		}
 
-// StartListening starts an event source
-func (el *EventListener) StartListening(ctx context.Context, dispatch func([]byte) error) error {
-	logger := logging.FromContext(ctx).
-		With(logging.LabelEventSourceType, el.GetEventSourceType(), logging.LabelEventName, el.GetEventName())
-	logger.Info("started processing the Github event source...")
+		ctx, cancel := context.WithCancel(ctx)
+		defer cancel()
 
-	githubEventSource := &el.GithubEventSource
+		f := func() {
+			for _, r := range githubEventSource.GetOwnedRepositories() {
+				for _, name := range r.Names {
+					hooks, _, err := router.githubClient.Repositories.ListHooks(ctx, r.Owner, name, nil)
+					if err != nil {
+						logger.Errorf("failed to list existing webhooks of %s/%s. err: %+v", r.Owner, name, err)
+						continue
+					}
+					h := getHook(hooks, formattedURL, githubEventSource.Events)
+					if h != nil {
+						router.hookIDs[r.Owner+","+name] = *h.ID
+						continue
+					}
+					logger.Infof("hook not found for %s/%s, creating ...", r.Owner, name)
+					h, _, err = router.githubClient.Repositories.CreateHook(ctx, r.Owner, name, hook)
+					if err != nil {
+						logger.Errorf("failed to create github webhook for %s/%s. err: %+v", r.Owner, name, err)
+						continue
+					}
+					router.hookIDs[r.Owner+","+name] = *h.ID
+					time.Sleep(500 * time.Millisecond)
+				}
+			}
+		}
 
-	route := webhook.NewRoute(githubEventSource.Webhook, logger, el.GetEventSourceName(), el.GetEventName(), el.Metrics)
+		// Github can not handle race condtions well - it might create multiple hooks with same config
+		// Randomly sleep some time to mitigate the issue.
+		s1 := rand.NewSource(time.Now().UnixNano())
+		r1 := rand.New(s1)
+		time.Sleep(time.Duration(r1.Intn(2000)) * time.Millisecond)
+		f()
+
+		go func() {
+			// Another kind of race conditions might happen when pods do rolling upgrade - new pod starts
+			// and old pod terminates, if DeleteHookOnFinish is true, the hook will be deleted from github.
+			// This is a workround to mitigate the race conditions.
+			logger.Info("starting github hooks manager daemon")
+			ticker := time.NewTicker(60 * time.Second)
```

Comment:
tickers leak references, I prefer to just put a sleep in the default branch

Reply:
Could you give more detail about that? I know `time.Tick()` "leaks", but never heard `NewTicker` has an issue.

But I did forget to add `defer ticker.Stop()`.

Reply:
Well there you go.

Reply:
Fixed all the same issues in the project.